doc/neps/nep-0021-advanced-indexing.rst

```diff
@@ -0,0 +1,537 @@
+advanced indexing inserts smaller dimensions.
+These examples may be hard to grasp without working knowledge of advanced
+indexing as of NumPy 1.9.
+
+Example array::
+
+    >>> arr = np.ones((5, 6, 7, 8))
+
+
+Legacy fancy indexing
+---------------------
+
+Note that the same result can be achieved with ``arr.lindex``, but the
+"future error" will still work in this case.
+
+Single index is transposed (this is the same for all indexing types)::
+
+    >>> arr[[0], ...].shape
+    (1, 6, 7, 8)
+    >>> arr[:, [0], ...].shape
+    (5, 1, 7, 8)
+
+
+Multiple indices are transposed *if* consecutive::
+
+    >>> arr[:, [0], [0], :].shape  # future error
+    (5, 1, 8)
+    >>> arr[:, [0], :, [0]].shape  # future error
+    (1, 5, 7)
+
+
+It is important to note that a scalar *is* integer array index in this sense
+(and gets broadcasted with the other advanced index)::
+
+    >>> arr[:, [0], 0, :].shape
+    (5, 1, 8)
+    >>> arr[:, [0], :, 0].shape  # future error (scalar is "fancy")
+    (1, 5, 7)
+
+
+Single boolean index can act on multiple dimensions (especially the whole
+array). It has to match (as of 1.10. a deprecation warning) the dimensions.
+The boolean index is otherwise identical to (multiple consecutive) integer
+array indices::
+
+    >>> # Create boolean index with one True value for the last two dimensions:
+    >>> bindx = np.zeros((7, 8), dtype=np.bool_)
+    >>> bindx[0, 0] = True
+    >>> arr[:, 0, bindx].shape
+    (5, 1)
+    >>> arr[0, :, bindx].shape
+    (1, 6)
+
+
+The combination with anything that is not a scalar is confusing, e.g.::
+
+    >>> arr[[0], :, bindx].shape  # bindx result broadcasts with [0]
+    (1, 6)
+    >>> arr[:, [0, 1], bindx].shape  # IndexError
+
+
+Outer indexing
+--------------
+
+Multiple indices are "orthogonal" and their result axes are inserted 
+at the same place (they are not broadcasted)::
+
+    >>> arr.oindex[:, [0], [0, 1], :].shape
+    (5, 1, 2, 8)
+    >>> arr.oindex[:, [0], :, [0, 1]].shape
+    (5, 1, 7, 2)
+    >>> arr.oindex[:, [0], 0, :].shape
+    (5, 1, 8)
+    >>> arr.oindex[:, [0], :, 0].shape
+    (5, 1, 7)
+
+
+Boolean indices results are always inserted where the index is::
+
+    >>> # Create boolean index with one True value for the last two dimensions:
+    >>> bindx = np.zeros((7, 8), dtype=np.bool_)
+    >>> bindx[0, 0] = True
+    >>> arr.oindex[:, 0, bindx].shape
+    (5, 1)
+    >>> arr.oindex[0, :, bindx].shape
+    (6, 1)
+
+
+Nothing changed in the presence of other advanced indices since::
+
+    >>> arr.oindex[[0], :, bindx].shape
+    (1, 6, 1)
+    >>> arr.oindex[:, [0, 1], bindx].shape
+    (5, 2, 1)
+
+
+Vectorized/inner indexing
+-------------------------
+
+Multiple indices are broadcasted and iterated as one like fancy indexing,
+but the new axes area always inserted at the front::
+
+    >>> arr.vindex[:, [0], [0, 1], :].shape
+    (2, 5, 8)
+    >>> arr.vindex[:, [0], :, [0, 1]].shape
+    (2, 5, 7)
+    >>> arr.vindex[:, [0], 0, :].shape
```

Comment:
I echo @jaimefrio's comment from way back that having `[0]` remove a dimension and `0` keep it is *extremely* confusing. But it may just mean I still do not understand fancy indexing...

Reply:
Frankly, I am a bit confused what you mean :). Any index "removes" the dimension in some sense (for vindex, oindex in some sense replaces it), the question is just where to put the new dims. Jaime had a comment suggesting to maybe keep the axis swapping in a less confusing manner then it used to exist.
Right now, I disagree with this. `vindex` should be as consistent as possible, there is nothing special enough about `[0]` compared to `[0, 1]`. There **might** be a "less confusing then currently fancy indexing like transposing" thing, but I personally believe you can use oindex then, it already does what you want, no?

Reply:
I just don't understand why here a leading `1` appears. But as said, my understanding of fancy indexing is quite hopeless.  I do agree with the always-transpose-to-the-front rule.

Reply:
Well, the thing is, there is a 1 sized dimension created by the  `[1]` so it has to go somewhere? And in vindex that somewhere is the front?

Reply:
Yes, that is logical, but then why is it not happening in the examples right above? Or are those mistakes?

Reply:
Well, its vindex, not oindex. All indexes (except slices I guess) are broadcast together for the result (just like legacy fancy indexing) and iterated together. So the extra dimension at the start has size 2 because it has the size of `[0]` and `[0, 1]` broadcast together.

Reply:
A, see, I knew I was missing something. Though this is particularly silly. Sorry for the noise.